plasmapy_nei/nei/nei.py

```diff
@@ -1,7 +1,1359 @@
+
+        new_T_e
+            The new electron temperature.
+
+        """
+
+        try:
+            index = self._index
+            elements = self.elements
+            self._time[index] = new_time
+            self._T_e[index] = new_T_e
+
+            for elem in elements:
+                self._ionic_fractions[elem][index, :] = new_ionfracs[elem][:]
+
+            # Calculate elemental and ionic number densities
+            n_elem = {elem: new_n * self.abundances[elem] for elem in elements}
+            number_densities = {
+                elem: n_elem[elem] * new_ionfracs[elem] for elem in elements
+            }
+
+            # Calculate the electron number density
+            n_e = 0.0 * u.cm ** -3
+            for elem in elements:
+                integer_charges = np.linspace(
+                    0, self.nstates[elem] - 1, self.nstates[elem]
+                )
+                n_e += np.sum(number_densities[elem] * integer_charges)
+
+            # Assign densities
+            self._n_e[index] = n_e
+            for elem in elements:
+                self._n_elem[elem][index] = n_elem[elem]
+                self._number_densities[elem][index, :] = number_densities[elem]
+
+        except Exception as exc:
+            raise NEIError(
+                f"Unable to assign parameters to Simulation instance "
+                f"for index {index} at time = {new_time}.  The "
+                f"parameters are new_n = {new_n}, new_T_e = {new_T_e}, "
+                f"and new_ionic_fractions = {new_ionfracs}."
+            ) from exc
+        finally:
+            self._index += 1
+
+    def _cleanup(self):
+        """
+        Clean up this class after the simulation is complete.
+
+        This method removes the excess elements from each array that
+        did not end up getting used for a time step in the simulation
+        and sets the ``last_step`` attribute.
+
+        """
+        nsteps = self._index
+
+        self._n_e = self._n_e[0:nsteps]
+        self._T_e = self._T_e[0:nsteps]
+        self._time = self._time[0:nsteps]
+
+        for element in self.elements:
+            self._ionic_fractions[element] = self._ionic_fractions[element][0:nsteps, :]
+            self._number_densities[element] = self._number_densities[element][
+                0:nsteps, :
+            ]
+
+        self._last_step = nsteps - 1
+
+        self._index = None
+
+    @property
+    def max_steps(self) -> int:
+        """
+        The maximum number of time steps allowed for this simulation.
+        """
+        return self._max_steps
+
+    @property
+    def last_step(self) -> int:
+        """The time index of the last step."""
+        return self._last_step
+
+    @property
+    def nstates(self) -> Dict[str, int]:
+        """
+        Return the dictionary containing atomic symbols as keys and the
+        number of ionic species for the corresponding element as the
+        value.
+        """
+        return self._nstates
```

Comment:
For this kind of field, if there's no setter anywhere... is this just Sphinx-driven development? :)